### PR TITLE
Add HTTPLocalRateLimitPolicy support

### PR DESCRIPTION
**Pull request by alpeb**
This adds the HTTPLocalRateLimitPolicy CRD, which is indexed by the policy controller and exposed by the inbound API.

- 81ebc08bd: HTTPLocalRateLimitPolicy CRD and related changes
- 01afd2304: policy controller central changes
- b09892529: rust tests updates and additions
- 2f455973c: golden files updates.

## Testing

In a cluster with linkerd and emojivoto injected, deploy these resources:

```yaml
 apiVersion: policy.linkerd.io/v1beta3
kind: Server
metadata:
  namespace: emojivoto
  name: web-http
spec:
  # permissive policy, so we don't require setting up authz
  accessPolicy: all-unauthenticated
  podSelector:
    matchLabels:
      app: web-svc
  port: http
  proxyProtocol: HTTP/1
```
```yaml
apiVersion: policy.linkerd.io/v1alpha1
kind: HTTPLocalRateLimitPolicy
metadata:
  namespace: emojivoto
  name: web-rl
spec:
  targetRef:
    group: policy.linkerd.io
    kind: Server
    name: web-http
  total:
    requestsPerSecond: 100
  identity:
    requestsPerSecond: 20
  overrides:
  - requestsPerSecond: 10
    clientRefs:
    - kind: ServiceAccount
      namespace: emojivoto
      name: default
```

```console
$ kubectl -n emojivoto get httplocalratelimitpolicies.policy.linkerd.io
NAME     TARGET_KIND   TARGET_NAME   TOTAL_RPS   IDENTITY_RPS
web-rl   Server        web-http      100         20
```

Then see how the RL policy is exposed at the inbound API under the protocol section, with `linkerd dg policy -n emojivoto po/web-85f6fb8564-jp67d 8080`:

```yaml
...
protocol:
  Kind:
    Http1:
      local_rate_limit:
        identity:
          requestsPerSecond: 20
        metadata:
          Kind:
            Resource:
              group: policy.linkerd.io
              kind: httplocalratelimitpolicy
              name: web-rl
        overrides:
        - clients:
            identities:
            - name: default.emojivoto.serviceaccount.identity.linkerd.cluster.local
          limit:
            requestsPerSecond: 10
        total:
          requestsPerSecond: 100
...
```

### CODE REVIEW

File: policy-test/tests/inbound_api.rs [hunk 1]
```diff
@@ -34,7 +34,10 @@ async fn server_with_server_authorization() {
         // that the update now uses this server, which has no authorizations
         let server = create(&client, mk_admin_server(&ns, "linkerd-admin", None)).await;
         let config = next_config(&mut rx).await;
-        assert_eq!(config.protocol, Some(grpc::defaults::proxy_protocol()));
+        assert_eq!(
+            config.protocol,
+            Some(grpc::defaults::proxy_protocol_no_ratelimit())
+        );
```
Comment (olix0r) [inline policy-test/tests/inbound_api.rs:40]: Preferable to not modify formatting of unrelated code -- can cause backport pain.

File: policy-test/src/grpc.rs [hunk 2]
```diff
@@ -376,24 +378,81 @@ impl hyper::service::Service<hyper::Request<tonic::body::BoxBody>> for GrpcHttp
...
+                    clients: Some(identities),
+                }
+            })
+            .collect();
+
+        inbound::HttpLocalRateLimit {
+            metadata: Some(Metadata {
+                kind: Some(metadata::Kind::Resource(Resource {
+                    group: "policy.linkerd.io".to_string(),
+                    kind: "httplocalratelimitpolicy".to_string(),
```
Comment (olix0r) [inline policy-test/src/grpc.rs:441]:
```suggestion
                    kind: "HTTPLocalRateLimitPolicy".to_string(),
```

File: policy-controller/k8s/index/src/inbound/index.rs [hunk 3]
```diff
@@ -1760,6 +1868,65 @@ impl PolicyIndex {
...
+        for (name, spec) in self.ratelimit_policies.iter() {
+            // Skip the policy if it doesn't apply to the server.
+            let ratelimit_policy::Target::Server(this_name) = &spec.target;
+            if this_name != server_name {
+                tracing::trace!(
+                    ns = %self.namespace,
+                    ratelimitpolicy = %name,
+                    server = %server_name,
+                    target = %name,
+                    "RateLimitPolicy does not target server",
```
Comment (olix0r) [inline policy-controller/k8s/index/src/inbound/index.rs:1881]:
```suggestion
                    "HTTPLocalRateLimitPolicy does not target server",
```

File: charts/linkerd-crds/templates/policy/http-local-ratelimit-policy.yaml [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  name: httplocalratelimitpolicies.policy.linkerd.io
+  annotations:
+    {{ include "partials.annotations.created-by" . }}
+  labels:
+    helm.sh/chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    linkerd.io/control-plane-ns: {{.Release.Namespace}}
+spec:
+  group: policy.linkerd.io
+  names:
+    kind: HTTPLocalRateLimitPolicy
```
Comment (olix0r) [inline charts/linkerd-crds/templates/policy/http-local-ratelimit-policy.yaml:14]: Do we need to define listKind (we do this for other types)?

```suggestion
    kind: HTTPLocalRateLimitPolicy
    listKind: HTTPLocalRateLimitPolicyList
```

File: policy-controller/grpc/src/inbound.rs [hunk 5]
```diff
@@ -327,3 +330,48 @@ fn to_authz(
         authentication: Some(authn),
     }
 }
+
+fn to_rate_limit(rl: &RateLimit) -> proto::HttpLocalRateLimit {
+    let meta = Metadata {
+        kind: Some(metadata::Kind::Resource(api::meta::Resource {
+            group: "policy.linkerd.io".to_string(),
+            kind: "httplocalratelimitpolicy".to_string(),
```
Comment (olix0r) [inline policy-controller/grpc/src/inbound.rs:338]:
```suggestion
            kind: "HTTPLocalRateLimitPolicy".to_string(),
```

File: policy-test/tests/inbound_api.rs (hunk 1 above)
Comment (alpeb) [reply policy-test/tests/inbound_api.rs:40]: It's not just formatting; `proxy_protocol_no_ratelimit` is a replacement for the `proxy_protocol` function that doesn't include rate-limiting. To minimize the change, I could instead keep `proxy_protocol` and create a new `proxy_protocol_with_ratelimiting`.

File: policy-controller/grpc/src/inbound.rs (hunk 5 above)
Comment (alpeb) [reply policy-controller/grpc/src/inbound.rs:338]: I was following the convention from the other Kinds in this file. Is it ok to deviate in this particular case given the length of this name?

File: charts/linkerd-crds/templates/policy/http-local-ratelimit-policy.yaml (hunk 4 above)
Comment (alpeb) [reply charts/linkerd-crds/templates/policy/http-local-ratelimit-policy.yaml:14]: Not currently used, but good to have :+1:

File: policy-controller/grpc/src/inbound.rs (hunk 5 above)
Comment (olix0r) [reply policy-controller/grpc/src/inbound.rs:338]: I think the conventions are mixed, but the most desirable outcome (from a readability and overall consistency point of view) is to use the canonical form with capitalization. Specifically we have simply pass-through the Kind for route resources:

https://github.com/linkerd/linkerd2/blob/8fb22d697d580c43dbed89a120b83246fb7d5437/policy-controller/grpc/src/inbound/grpc.rs#L56-L63
https://github.com/linkerd/linkerd2/blob/8fb22d697d580c43dbed89a120b83246fb7d5437/policy-controller/grpc/src/inbound/http.rs#L56-L61
https://github.com/linkerd/linkerd2/blob/8fb22d697d580c43dbed89a120b83246fb7d5437/policy-controller/k8s/index/src/inbound/index.rs#L1837

File: Cargo.toml [hunk 6]
```diff
@@ -14,5 +14,4 @@ members = [
 lto = "thin"
 
 [patch.crates-io]
-# TODO(Zahari): switch released version once TLS protocol support is out.
-linkerd2-proxy-api = { git = 'https://github.com/linkerd/linkerd2-proxy-api', rev = 'c5648ae2a1e405cc6b8aca20522356ebdf20f1ea' }
\ No newline at end of file
+linkerd2-proxy-api = { git = "https://github.com/linkerd/linkerd2-proxy-api.git", branch = "alpeb/rate-limiting" }
```
Comment (olix0r) [inline Cargo.toml:17]: Pulled out the API bump into a separate PR https://github.com/linkerd/linkerd2/pull/13290

File: policy-controller/k8s/index/src/inbound/index.rs [hunk 7]
```diff
@@ -1760,6 +1868,65 @@ impl PolicyIndex {
         authzs
     }
 
+    fn client_ratelimit(&self, server_name: &str) -> Option<RateLimit> {
+        for (name, spec) in self.ratelimit_policies.iter() {
```
Comment (olix0r) [inline policy-controller/k8s/index/src/inbound/index.rs:1872]: This seems like it might be non-deterministic if there are multiple policies targetting a server. Since we're iterating over a HashMap, we have no guarantees about the iteration order.

Are there guarantees elsewhere that there is only one policy for a given server? Or how do we deal with ambiguity?

Comment (olix0r) [reply policy-controller/k8s/index/src/inbound/index.rs:1872]: Ah, I don't think that's sufficient: we can't rely on admission controllers for any inter-resource validation constraints (like uniqueness). When applying multiple resources, for instances, the validator's view of the cluster may lag the reality. Or the validator may not be invoked for certain resources--resources may be applied when the validator was down or the config was not valid. Admission controllers are only useful to validate that a resource is _internally_ consistent--i.e. to apply schema validation.

This is why we use statuses (e.g. for HTTPRoutes) to control 'acceptance'--we must accept that resources may be created in the cluster, and we must have an eventually consistent view of those that are actually admitted by the controller.

Comment (olix0r) [reply policy-controller/k8s/index/src/inbound/index.rs:1872]: I expect that the best path forward is:

1. In this PR, we should determine/document precedence for these policies (probably using the same rules for HTTPRoute attachment). If the controller honors these precedence rules, then we can at least ensure that there is predictable/consistent behavior.
2. We should follow up by adding statuses to the policy resources that communicate the acceptance status of the resource.

Comment (alpeb) [reply policy-controller/k8s/index/src/inbound/index.rs]: Ok in 21e74d633 I've changed this particular function to only use the first RL by creation time and name. This means we keep the model to be one-to-one between servers and RLs, and we're not combining rules in any way.

File: policy-controller/k8s/index/src/inbound/index.rs [hunk 8]
```diff
@@ -1760,6 +1868,80 @@ impl PolicyIndex {
...
+                identity: spec.identity.as_ref().map(|l| Limit {
+                    requests_per_second: l.requests_per_second,
+                }),
+                overrides,
+            });
+        }
+
+        None
+    }
+
```
Comment (olix0r) [inline policy-controller/k8s/index/src/inbound/index.rs:1944]: A suggestion:

If we filter the policies first, the sort will only be over matching policies. In the expected case where there is only one rate limiting policy, this should make the sort_by free, even when there are multiple rate limiting policies in a namespace.

That simplification also allows us to early return and eliminate a for/continue loop.

```suggestion
    fn client_ratelimit(&self, server_name: &str) -> Option<RateLimit> {
        use ratelimit_policy::{ClientRef, Target};

        // sort the ratelimit policies by creation timestamp and name so we can
        // deterministically always return the same policy when more than one
        // point to the same server
        let mut rate_limits = self
            .ratelimit_policies
            .iter()
            .filter(|(_, spec)| matches!(spec.target, Target::Server(ref n) if n == server_name))
            .collect::<Vec<_>>();
        rate_limits.sort_by(|(a_name, a), (b_name, b)| {
            let by_ts = match (&a.creation_timestamp, &b.creation_timestamp) {
                (Some(a_ts), Some(b_ts)) => a_ts.cmp(b_ts),
                (None, None) => std::cmp::Ordering::Equal,
                // entries with timestamps are preferred over ones without
                (Some(_), None) => return std::cmp::Ordering::Less,
                (None, Some(_)) => return std::cmp::Ordering::Greater,
            };
            by_ts.then_with(|| a_name.cmp(b_name))
        });

        let (name, spec) = rate_limits.first()?;
        tracing::trace!(
            ns = %self.namespace,
            ratelimitpolicy = %name,
            server = %server_name,
            "HTTPLocalRateLimitPolicy targets server",
        );

        let overrides = spec
            .overrides
            .iter()
            .map(|ovr| {
                let client_identities = ovr
                    .client_refs
                    .iter()
                    .map(|ClientRef::ServiceAccount { namespace, name }| {
                        let namespace = namespace.as_deref().unwrap_or(&self.namespace);
                        self.cluster_info.service_account_identity(namespace, name)
                    })
                    .collect();

                Override {
                    requests_per_second: ovr.requests_per_second,
                    client_identities,
                }
            })
            .collect();

        Some(RateLimit {
            name: name.to_string(),
            total: spec.total.as_ref().map(|l| Limit {
                requests_per_second: l.requests_per_second,
            }),
            identity: spec.identity.as_ref().map(|l| Limit {
                requests_per_second: l.requests_per_second,
            }),
            overrides,
        })
    }
```

**Review by olix0r — APPROVED** [review]
